src/main/java/com/cleanroommc/groovyscript/compat/mods/astralsorcery/Constellation.java

```diff
@@ -0,0 +1,248 @@
+
+        private String name;
+        private Color color = null;
+        private ConstellationBuilder.Type type;
+        private final ArrayList<Tuple<Point,Point>> connections = new ArrayList<>();
+        private final ArrayList<MoonPhase> phases = new ArrayList<>();
+
+        private enum Type {
+            MAJOR,
+            MINOR,
+            WEAK
+        }
+
+        public ConstellationBuilder name(String name) {
+            this.name = name;
+            return this;
+        }
+
+        public ConstellationBuilder color(String color) {
+            this.color = Color.decode(color);
+            return this;
+        }
+
+        public ConstellationBuilder major() {
+            this.type = Type.MAJOR;
+            return this;
+        }
+
+        public ConstellationBuilder minor() {
+            this.type = Type.MINOR;
+            return this;
+        }
+
+        public ConstellationBuilder weak() {
+            this.type = Type.WEAK;
+            return this;
+        }
+
+        public ConstellationBuilder connection(int x1, int y1, int x2, int y2) {
+            Point s1 = new Point(x1,y1);
+            Point s2 = new Point(x2,y2);
+            Tuple<Point, Point> line = new Tuple<>(s1,s2);
+            Tuple<Point, Point> lineReverse = new Tuple<>(s2,s1);
+            if (!connections.contains(line) && !connections.contains(lineReverse)) {
+                connections.add(line);
+            }
+            return this;
+        }
+
+        public ConstellationBuilder phase(MoonPhase moonPhase) {
+            this.phases.add(moonPhase);
+            return this;
+        }
+
+        public ConstellationBuilder phase(MoonPhase... moonPhases) {
+            this.phases.addAll(Arrays.asList(moonPhases));
+            return this;
+        }
+
+        public ConstellationBuilder phase(Collection<MoonPhase> moonPhases) {
+            this.phases.addAll(moonPhases);
+            return this;
+        }
+
+        private boolean validate() {
+            if (this.name == null || this.name.equals("")) {
+                GroovyLog.msg("Error adding Astral Sorcery Constellation: ")
+                        .add("name must be provided")
+                        .error()
+                        .post();
+                return false;
+            }
+            if (this.connections.equals(new ArrayList<>())) {
+                GroovyLog.msg("Error adding Astral Sorcery Constellation: ")
+                        .add("connections must not be empty")
+                        .error()
+                        .post();
+                return false;
+            }
+            if (this.type.equals(Type.MINOR) && this.phases.size() == 0) {
+                GroovyLog.msg("Error adding Astral Sorcery Constellation: ")
+                        .add("minor constellations require at least one moon phase")
+                        .error()
+                        .post();
+                return false;
+            }
```

Comment:
While this is not wrong, it is not optimal. Assume someone is writing a recipe which unintentionally match all 3 cases. First it checks for the name. The recipe will not be registered and a message to the log will appear. The users fixes it and reloads. Now it runs into the 2nd error. The same thing happens again for this and the 3rd error. The user had to reload 3 times to make the recipe work.
It would be much better if all 3 errors are logged at the same time. You can look at the AbstractRecipeBuilder to see how i did it.